R/parse.R

```diff
@@ -118,6 +123,52 @@ coerce_datetime <- function(x, to, ...) {
   }
 }
 
+# Parse character dates received from Connect which use RFC 3339.
+#
+# R parses as ISO 8601. When specifying %z, it expects time zones to be
+# specified as `-1400` to `+1400`.
+# Connect returns times in RFC 3339. It denotes time zones with `-14:00` to
+# `+14:00`, and indicates zero offset with `Z`.
+# https://github.com/golang/go/blob/54fe0fd43fcf8609666c16ae6d15ed92873b1564/src/time/format.go#L86
+# I don't understand why replacing Z with `+0000` and parsing with %z doesn't work, but it doesn't.
+# We have to parse zero-offset time stamps expecting literal Zs.
+# For example:
+# - "2023-08-22T14:13:14Z"
+# - "2023-08-22T15:13:14+01:00"
+# - "2020-01-01T00:02:03-01:00"
+parse_connect_rfc3339 <- function(x) {
+  # Convert any timestamps with offsets to a format recognized by `strptime`.
+  x <- gsub("([+-]\\d\\d):(\\d\\d)$", "\\1\\2", x)
+  
+  # Times with and without offsets require different formats, so create a vector
+  # of formats to be used in parallel.
+  format_strings <- ifelse(
+    grepl("Z$", x),
+    "%Y-%m-%dT%H:%M:%SZ",
+    "%Y-%m-%dT%H:%M:%S%z"
+  )
+
+  # Parse with an inner call to `strptime()`; convert the resulting `POSIXlt`
+  # object to `POSIXct`.
+  #
+  # We must specify `tz` in the inner call to correctly compute date math.
+  # Specifying `tz` when parsing just changes the time zone without doing any
+  # date math!
+  #
+  # > xlt
+  # [1] "2024-08-29 16:36:33 EDT"
+  # > tzone(xlt)
+  # [1] "America/New_York"
+  # > as.POSIXct(xlt, tz = "UTC")
+  # [1] "2024-08-29 16:36:33 UTC"
+  #
+  # `purrr::map2_vec()` converts to POSIXct automatically, but we need
+  # `as.POSIXct()` in there to account for only one item.
+  purrr::map2_vec(x, format_strings, function(.x, .y) {
```

Comment:
Is the `function(.x, .y)` called on each element? That gives you the option of moving format string selection into this function call rather than computing it separately.

I'm fine leaving this as-is; your choice about which option is more clear.

Reply:
Is it actually possible for the format string to vary by element (e.g., a creation timestamp in user records would be omitted as UTC for some users and local time for others)? My understanding of the issue was that it was about the configuration of the Connect server.

If it is the case that the Connect server is consistent with itself, then we can determine the format string to use from the first element, and we won't need purrr at all down here, just strptime.

Reply:
@aronatkins Yes, it is called on each element. In that case, I could just use `map_vec` and not bother with the parallel creation.

@nealrichardson Yeah, I think you're right, I don't think that a given Connect server will emit multiple records with different time zone offsets, so we can probably determine the format with the first element and use a single format specification.

Reply:
@nealrichardson I think I would prefer to keep a function such that it's vectorized in a way that can parse an character of strings that are in the assumed format, rather than introduce the additional assumption that the first element's time zone offset is representative of all of the elements. I know this function is a single-purpose one, but we don't have anything in place that's guaranteeing that all the time zone offsets in the incoming string are the same (other than, like, we write the Connect server).

I guess I'd rather write something that's robust to misuse or unexpected edge cases than something which is marginally more performant but more brittle.

Reply:
> we don't have anything in place that's guaranteeing that all the time zone offsets in the incoming string are the same (other than, like, we write the Connect server)

You're right that the performance impact is negligible--any time we're fetching things over HTTP, it's not worth worrying about microseconds of optimization in the client code.

I would counter, as a general rule, that we should avoid the tendency of overgeneralizing and flexibilizing our client code. There's already enough flex we have to add to support different versions of Connect, etc., we don't need more. This to me feels like it should be a design constraint in the Connect API: all timestamps are emitted the same way. (Ideally, Connect would be even stricter, always emitting UTC, ISO 8601, etc.) And thus API clients can assume that timestamps are a certain way. 

In practice here, it's not a big deal--the code that makes this flexible is probably simpler. Just wanted to make the more general design point.

Reply:
Thanks -- I definitely understand the need to not over-generalize, and it's good to keep that in mind. I think this code is striking the right balance — it's pretty narrow in what it expects (only parsing timestamps in the subset of RFC 3339 which includes time zone offsets ending in `Z` / `±hh:mm`).

It's a moot point now, but I realize that another reason constraining it further feels wrong is that it breaks with expectations of how vectorization works in R: functions that operate on vectors generally don't alter processing of subsequent items based on the first item. Better to not break those assumptions IMO.